sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/CanonicalizeSuite.scala

```diff
@@ -50,4 +59,44 @@ class CanonicalizeSuite extends SparkFunSuite {
     assert(range.where(arrays1).sameResult(range.where(arrays2)))
     assert(!range.where(arrays1).sameResult(range.where(arrays3)))
   }
+
+  test("SPARK-26402: GetStructField with different optional names are semantically equal") {
+    val expId = NamedExpression.newExprId
+    val qualifier = Seq.empty[String]
+    val structType = StructType(
+      StructField("a", StructType(StructField("b", IntegerType, false) :: Nil), false) :: Nil)
+
+    val fieldB1 = GetStructField(
+      AttributeReference("data1", structType, false)(expId, qualifier),
+      0, Some("b1"))
+    val fieldB2 = GetStructField(
+      AttributeReference("data2", structType, false)(expId, qualifier),
+      0, Some("b2"))
+    assert(fieldB1.semanticEquals(fieldB2))
+
+    val fieldA1 = GetStructField(
+      GetStructField(
+        AttributeReference("data1", structType, false)(expId, qualifier),
+        0, Some("a1")),
+      0, Some("b1"))
+    val fieldA2 = GetStructField(
+      GetStructField(
+        AttributeReference("data2", structType, false)(expId, qualifier),
+        0, Some("a2")),
+      0, Some("b2"))
+    assert(fieldA1.semanticEquals(fieldA2))
+
+    // End-to-end test case
+    val testRelation = LocalRelation('a.int)
```

Comment:
This is not a real end-to-end test...

How about add the following test to SQLQuerySuite?
```
sql("create table t (s struct<i: Int>) using json")
sql("select s.I from t group by s.i")
```
currently it fials with
```
org.apache.spark.sql.AnalysisException: expression 'default.t.`s`' is neither present in the group by, nor is it an aggregate function
```

Reply:
This one makes sense, and is addressed by this PR.

Reply:
then can we remove this part? i.e. code between L89 to L99

Reply:
+1

Reply:
This test can be a part of `BinaryComparisonSimplificationSuite` for `SimplifyBinaryComparison`. So far, there is no test case for `struct` type. Since this PR, `SimplifyBinaryComparison` can remove `s.I <=> s.i` as we see here.

Reply:
I'm curious that is that removed too when case sensitive mode is turned on?

Reply:
It will fail at name resolution.

Reply:
@viirya I added a test to show in case insensitive mode, it will fail.